pkg/sql/schemaexpr/column.go

```diff
@@ -163,3 +123,128 @@ func FormatColumnForDisplay(
 	}
 	return f.CloseAndGetString(), nil
 }
+
+// iterColDescriptors iterates over the expression's variable columns and
+// calls f on each.
+//
+// If the expression references a column that does not exist in the table
+// descriptor, iterColDescriptors errs with pgcode.UndefinedColumn.
+func iterColDescriptors(
+	desc *sqlbase.MutableTableDescriptor, rootExpr tree.Expr, f func(*sqlbase.ColumnDescriptor) error,
+) error {
+	_, err := tree.SimpleVisit(rootExpr, func(expr tree.Expr) (recurse bool, newExpr tree.Expr, err error) {
+		vBase, ok := expr.(tree.VarName)
+		if !ok {
+			// Not a VarName, don't do anything to this node.
+			return true, expr, nil
+		}
+
+		v, err := vBase.NormalizeVarName()
+		if err != nil {
+			return false, nil, err
+		}
+
+		c, ok := v.(*tree.ColumnItem)
+		if !ok {
+			return true, expr, nil
+		}
+
+		col, dropped, err := desc.FindColumnByName(c.ColumnName)
+		if err != nil || dropped {
+			return false, nil, pgerror.Newf(pgcode.UndefinedColumn,
+				"column %q does not exist, referenced in %q", c.ColumnName, rootExpr.String())
+		}
+
+		if err := f(col); err != nil {
+			return false, nil, err
+		}
+		return false, expr, err
+	})
+
+	return err
+}
+
+// dummyColumn represents a variable column that can type-checked. It is used
+// in validating check constraint and partial index predicate expressions. This
+// validation requires that the expression can be both both typed-checked and
+// examined for variable expressions.
+type dummyColumn struct {
+	typ *types.T
+	// name is only used for error-reporting.
```

Comment:
This comment is out of date now -- see the comments from where it was moved from in `sqlbase`

Reply:
Done.